linkerd/app/core/src/classify.rs

```diff
@@ -71,17 +71,16 @@ impl classify::Classify for Request {
         match self {
             Self::Profile(classes) => Response::Profile(classes.clone()),
 
-            Self::ClientPolicy(ClientPolicy::Http(policy)) => Response::Http(policy.clone()),
+            Self::ClientPolicy(ClientPolicy::Http(policy)) => {
+                if is_grpc(req.headers()) {
+                    return Response::Grpc(Default::default());
+                }
+                Response::Http(policy.clone())
+            }
             Self::ClientPolicy(ClientPolicy::Grpc(policy)) => Response::Grpc(policy.clone()),
 
             Self::Default => {
-                let is_grpc = req
-                    .headers()
-                    .get(http::header::CONTENT_TYPE)
-                    .and_then(|v| v.to_str().ok())
-                    .map(|ct| ct.starts_with("application/grpc"))
-                    .unwrap_or(false);
-                if is_grpc {
+                if is_grpc(req.headers()) {
                     return Response::Grpc(Default::default());
                 }
                 Response::default()
```

Comment:
nit, take it or leave it: it occurs to me that we don't actually need to repeat this branch twice, and we could instead write
```rust
        match self {
            Self::Profile(classes) => Response::Profile(classes.clone()),
            Self::ClientPolicy(ClientPolicy::Grpc(policy)) => Response::Grpc(policy.clone()),
            // If the request's content type indicates that it is gRPC, use the
            // default gRPC classifier rather than a configured HTTP classifier.
            _ if is_grpc(req.headers()) => Response::Grpc(Default::default()),
            Self::ClientPolicy(ClientPolicy::Http(policy)) => Response::Http(policy.clone()),
            Self::Default => Response::default(),
        }
```
but it's up to you whether you think that's easier to follow or not...